crates/dslab-mp/src/mc/model_checker.rs

```diff
@@ -73,43 +66,64 @@ impl ModelChecker {
 
         Self {
             system: McSystem::new(nodes, mc_net, events, trace_handler),
-            strategy,
         }
     }
 
+    fn run_impl<F, S>(&mut self, strategy: &mut S, preliminary_callback: F) -> McResult
+    where
+        F: FnOnce(&mut McSystem),
+        S: Strategy,
+    {
+        self.system.trace_handler.borrow_mut().push(LogEntry::McStarted {});
+        preliminary_callback(&mut self.system);
+        strategy.mark_visited(self.system.get_state());
+        strategy.run(&mut self.system)
+    }
+
     /// Runs model checking and returns the result on completion.
-    pub fn run(&mut self) -> McResult {
-        self.strategy.mark_visited(self.system.get_state());
-        self.strategy.run(&mut self.system)
+    pub fn run<S: Strategy>(&mut self, strategy_config: StrategyConfig) -> McResult {
+        self.run_with_change::<_, S>(strategy_config, |_| {})
     }
 
     /// Runs model checking after applying callback.
-    pub fn run_with_change<F>(&mut self, preliminary_callback: F) -> McResult
+    pub fn run_with_change<F, S>(&mut self, strategy_config: StrategyConfig, preliminary_callback: F) -> McResult
```

Comment:
От `F` здесь и в `run_from_states_with_change` можно избавиться так:

```suggestion
    pub fn run_with_change<S>(&mut self, strategy_config: StrategyConfig, preliminary_callback: impl FnOnce(&mut McSystem)) -> McResult
```

Reply:
Сделал так, да